### PR TITLE
Fix scheduling of ClusterInfoService#refresh

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the `InternalClusterInfoService` uses the
`LocalNodeMasterListener` interface to start/stop its operations. Since
the `onMaster` and `offMaster` methods are called on the `MANAGEMENT`
threadpool, there's no guarantee that they run in the correct sequence,
which could result in an elected master failing to regularly update the
cluster info.

Since this service is also a `ClusterStateListener` we may as well drop
the usage of the `LocalNodeMasterListener` interface and simply update
the status of the local node on the applier thread in `clusterChanged`
to ensure consistency.

Additionally, today the `InternalClusterInfoService` uses a simple flag
to track whether the local node is the elected master or not. If the
node stops being the master and then starts again within a few seconds
then the scheduled updates from the old mastership might carry on
running in addition to the ones for the new mastership.

This commit addresses that by tracking the identity of the scheduled
update job and creating a new job for each mastership.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 1]
```diff
@@ -127,75 +131,53 @@ void setUpdateFrequency(TimeValue updateFrequency) {
...
-            if (clusterService.state().getNodes().getDataNodes().size() > 1) {
-                // Submit an info update job to be run immediately
-                threadPool.executor(executorName()).execute(this::maybeRefresh);
-            }
-        } catch (EsRejectedExecutionException ex) {
-            logger.debug("Couldn't schedule cluster info update task - node might be shutting down", ex);
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster() && refreshAndRescheduleRunnable.get() == null) {
+            logger.trace("elected as master, scheduling cluster info update tasks");
+            executeRefresh(clusterService.state(), "became master");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:137]: Maybe just `event.state()` here?

**Review by ywelsch — APPROVED** [review]
It's annoying that this service is blocking a management thread (and there is no reason for it to do so). Anyway, removing that would be a larger change. Getting the data race fixed is a good step.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: ++ 6209182

**Comment by DaveCTurner** [comment]
Yeah the blocking nature of `refresh()` is indeed unnecessary but fixing that is orthogonal to this.

**Pull request merged**